File: src/main/java/usach/tingeso/services/ReportService.java
package usach.tingeso.services;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import usach.tingeso.entities.ReportEntity;
import usach.tingeso.repositories.ReportRepository;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class ReportService {
    @Autowired
    ReportRepository reportRepository;
    @Autowired
    TicketService ticketService;
    @Autowired
    private RepairService repairService;

    // Basic CRUD operations
    public ReportEntity getReportById(Long id){
        return reportRepository.findById(id).orElse(null);
    }

    public ReportEntity saveReport(ReportEntity report){
        if (report == null)
            return null;
        return reportRepository.save(report);
    }

    public void deleteReport(Long id){
        if (reportRepository.existsById(id)) {
            reportRepository.deleteById(id);
        }
    }

    public ReportEntity updateReport(ReportEntity report){
        if (reportRepository.existsById(report.getIdReport())) {
            return reportRepository.save(report);
        }
        return null;
    }

    // Here starts the code with Rx


    // R1
    public ReportEntity saveR1(ReportEntity report){
        List<String> allTicketValues = ticketService.getAllTicketValues();
        if (allTicketValues != null) {
            String r1Details = String.join(",\n ", allTicketValues);
            report.setR1Details(r1Details);
        }
        return reportRepository.save(report);
    }

    // R2
    public ReportEntity saveR2(ReportEntity report){
        String allTicketValues = repairService.getRepairTypeCountAndTotalPrice();
        if (allTicketValues != null) {
            String r2Details = String.join(",\n ", allTicketValues);
            report.setR2RepairsVsVehiclesByTotalType(r2Details);
        }
        return reportRepository.save(report);
    }

    // R3
    public ReportEntity saveR3(ReportEntity reporte){
        String allTicketValues = repairService.getAverageTotalRepairAmountByBrand();
        if (allTicketValues != null) {
            String r3Details = String.join(",\n ", allTicketValues);
            reporte.setR3AverageByBrand(r3Details);
        }
        return reportRepository.save(reporte);
    }
    // R4
    public ReportEntity saveR4(ReportEntity report){
        String allTicketValues = repairService.getRepairTypeMotorTypeCountAndTotalPrice();
        if (allTicketValues != null) {
            String r4Details = String.join(",\n ", allTicketValues);
            report.setR4RepairsVsVehiclesByTotalEngine(r4Details);
        }
        return reportRepository.save(report);
    }

}

File: pep1-frontend/pep1/src/services/report.service.js
import axios from 'axios';

const API_URL = 'http://localhost:8090/api/v1/report';

function getReportById(id) {
    return axios.get(`${API_URL}/${id}`);
}

function saveReport(report) {
    return axios.post(`${API_URL}/`, report);
}

function deleteReport(id) {
    return axios.delete(`${API_URL}/${id}`);
}

function updateReport(report) {
    return axios.put(`${API_URL}/`, report);
}

function saveR1(report) {
    return axios.post(`${API_URL}/r1`, report);
}

function saveR2(report) {
    return axios.post(`${API_URL}/r2`, report);
}

function saveR3(report) {
    return axios.post(`${API_URL}/r3`, report);
}

function saveR4(report) {
    return axios.post(`${API_URL}/r4`, report);
}

function saveInit(report) {
    return axios.put(`${API_URL}/init`, report);
}

export default {
    getReportById,
    saveReport,
    deleteReport,
    updateReport,
    saveR1,
    saveR2,
    saveR3,
    saveR4,
    saveInit
}